common/preconditioner/isai_kernels.hpp.inc

```diff
@@ -0,0 +1,220 @@
+    IndexType num_rows, const IndexType *__restrict__ m_row_ptrs,
+    const IndexType *__restrict__ m_col_idxs,
+    const ValueType *__restrict__ m_values, IndexType *__restrict__ i_row_ptrs,
+    IndexType *__restrict__ i_col_idxs, ValueType *__restrict__ i_values,
+    Callable trs_solve)
+{
+    static_assert((subwarp_size & (subwarp_size - 1)) == 0,
+                  "subwarp_size must be a power of 2");
+
+    const IndexType gidx = thread::get_thread_id_flat<IndexType>();
+    const IndexType row = gidx / subwarp_size;
+    const int tidx = threadIdx.x;
+
+    auto this_block = group::this_thread_block();
+    auto subwarp = group::tiled_partition<subwarp_size>(this_block);
+    const int subwarp_id = subwarp.thread_rank();
+
+    // Here, we use an intermediate shared memory block with a primitive
+    // datatype to produce no warning in case of complex types.
+    // The issue with thrust::complex is that it has a non-empty default
+    // constructor, which apparently does not work with non-dynamic shared
+    // memory
+    __shared__ char shared_memory[sizeof(ValueType) * subwarp_size *
+                                  subwarp_size * subwarps_per_block];
+    auto trisystem = reinterpret_cast<ValueType *>(shared_memory);
+
+    const int end_trisystem = blockDim.x * subwarp_size;
+    const int trisystem_start =
+        (tidx / subwarp_size) * subwarp_size * subwarp_size;
+
+    // Allow at most subwarp_size * subwarp_size elements in a subwarp trisystem
+    for (int i = 0; i < subwarp_size; ++i) {
+        trisystem[trisystem_start + subwarp_id + i * subwarp_size] =
+            zero<ValueType>();
+    }
+    this_block.sync();
+
+    if (row >= num_rows) {
+        return;
+    }
+
+    const auto i_row_begin = i_row_ptrs[row];
+    const auto i_row_end = i_row_ptrs[row + 1] - i_row_begin < subwarp_size
+                               ? i_row_ptrs[row + 1]
+                               : i_row_begin + subwarp_size;
+    const auto num_row_elems = i_row_end - i_row_begin;
+
+    if (subwarp_id < num_row_elems) {
+        const auto col = i_col_idxs[i_row_begin + subwarp_id];
+        const auto m_row_end = m_row_ptrs[col + 1];
+        auto m_row_ptr = m_row_ptrs[col];
+        auto i_row_ptr = i_row_begin;
+        int idx{trisystem_start +
+                subwarp_id};  // write to trisystem in row major
+
+        while (m_row_ptr < m_row_end && i_row_ptr < i_row_end &&
+               idx < end_trisystem) {
+            const auto sparsity_col = i_col_idxs[i_row_ptr];
+            const auto m_col = m_col_idxs[m_row_ptr];
+            if (sparsity_col == m_col) {
+                trisystem[idx] = m_values[m_row_ptr];
+                idx += subwarp_size;
+                ++m_row_ptr;
+                ++i_row_ptr;
+            } else if (m_col < sparsity_col) {
+                ++m_row_ptr;
+            } else {
+                ++i_row_ptr;
+                idx += subwarp_size;
+            }
+        }
+    }
```

Comment:
I tried to use the `sequential_merge` function, however, I have some issues with it:
- I need to protect `trisystem` from out-of-bound access (otherwise, the whole kernel could crash)
- the Index where I want to write to increases by `subwarp_size`, and not by 1 (which the `sequential_merge` uses)

The following code does work and fulfills all criteria I have.
However, it is likely slower than the original code because it
- does not cancel the whole merge if `idx` is too high
- creates a lot of unnecessary temporaries (minor issue)
- uses a non-const lambda where I am not sure how good the compiler deals with it

Since the lambda merge function needs to take care of some edge cases (and ensure that `idx` has the proper value), it is also less descriptive in my opinion, so I would prefer to keep the current implementation.
```c++
sequential_merge(m_col_idxs + m_row_ptr, m_row_end - m_row_ptr,
                 i_col_idxs + i_row_ptr, i_row_end - i_row_ptr,
                 [idx, m_values, m_row_ptr, end_trisystem](
                     IndexType m_idx, IndexType m_col, IndexType i_idx,
                     IndexType i_col, IndexType ) mutable {
                     if (idx < end_trisystem) {
                         if (m_col == i_col) {
                             trisystem[idx] = m_values[m_row_ptr + m_idx];
                             idx += subwarp_size;
                         } else if (i_col < m_col) {
                             idx += subwarp_size;
                         }
                     }
                 });
```

Reply:
This is roughly what I meant, yes. I think you can address most of the points you mentioned by noting that every time i_idx gets incremented, idx gets increased by subwarp_size, and taking the captured variables by reference (since the lambda doesn't leave the scope):
```cpp
sequential_merge(m_col_idxs + m_row_ptr, m_row_end - m_row_ptr,
                 i_col_idxs + i_row_ptr, i_row_end - i_row_ptr,
                 [&](IndexType m_idx, IndexType m_col, IndexType i_idx,
                     IndexType i_col, IndexType) {
                     auto idx = subwarp_size * i_idx + idx_base // == trisystem_start + local_id
                     if (idx < end_trisystem && m_col == i_col) {
                             trisystem[idx] = m_values[m_row_ptr + m_idx];
                     }
                 });
```
The remaining issue I see is the early exit for `idx >= end_trisystem`. We could add a bool return value to the callback (we might actually be able to use that elsewhere as well, now that I think of it), but that might complicate other parts of the code where sequential_merge is used. What do you think - Is this version of the code sufficiently descriptive? Or would you prefer to leave it as it is?

Reply:
How the change for early exit could look:
```
--- a/common/components/merging.hpp.inc
+++ b/common/components/merging.hpp.inc
@@ -220,7 +223,10 @@ __device__ void sequential_merge(const ValueType *a, IndexType a_size,
     auto a_cur = detail::checked_load(a, a_begin, a_size);
     auto b_cur = detail::checked_load(b, b_begin, b_size);
     while (c_begin < c_size) {
-        merge_fn(a_begin, a_cur, b_begin, b_cur, c_begin);
+        bool cont = merge_fn(a_begin, a_cur, b_begin, b_cur, c_begin);
+        if (!cont) {
+            break;
+        }
         auto a_advance = a_cur < b_cur;
         auto b_advance = !a_advance;
         a_begin += a_advance;
```

Reply:
I would personally prefer to leave it as it is if that is fine with you. Your code does work (and is a lot more elegant than the `mutable` version I had), however, it is still less descriptive **to me** than just the plain code.